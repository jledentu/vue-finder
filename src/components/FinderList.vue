<script>
import FinderItem from "./FinderItem";
import FinderListDropZone from "./FinderListDropZone";

function renderItems(h, { props, listeners }) {
  const DropZoneComponent = props.dropZoneComponent;

  function onItemEvent(eventName, itemId) {
    return event => {
      listeners[eventName](itemId, event);
    };
  }

  function onDragStart(item) {
    return event => {
      event.dataTransfer.setData("id", item.id);
      listeners["item-dragged"](item.id, event);
    };
  }

  function onDragOver(item) {
    return event => {
      if (item.children && item.children.length) {
        listeners["item-expanded"](item.id, event);
      }
    };
  }

  return props.items.map(item => [
    ...[
      props.dragEnabled && <DropZoneComponent key={`drop-zone-${item.id}`} />
    ],

    <FinderItem
      key={`item-${item.id}`}
      node={item}
      treeModel={props.treeModel}
      class={{ draggable: props.dragEnabled }}
      selectable={props.selectable}
      enabled={props.dragEnabled}
      dragged={props.draggedItem.id === item.id}
      vOn:start={onDragStart(item)}
      vOn:dragover_native={onDragOver(item)}
      vOn:drop_native={onItemEvent("drop", item.id)}
      vOn:dragend_native={onItemEvent("dragend", item.id)}
    >
      {item.label}
    </FinderItem>
  ]);
}

export default {
  name: "FinderList",
  functional: true,
  props: {
    items: {
      type: Array,
      default: () => []
    },
    itemComponent: {
      type: Object,
      default: () => FinderItem
    },
    dropZoneComponent: {
      type: Object,
      default: () => FinderListDropZone
    },
    treeModel: {
      type: Object,
      required: true
    },
    selectable: {
      type: Boolean,
      default: false
    },
    dragEnabled: {
      type: Boolean,
      default: Boolean
    },
    draggedItem: {
      type: Object,
      default: () => ({})
    }
  },
  render(h, { props, listeners }) {
    const DropZoneComponent = props.dropZoneComponent;

    return [
      <div class="list">
        {[
          ...renderItems(h, { props, listeners }),
          ...[props.dragEnabled && <DropZoneComponent class="last" />]
        ]}
      </div>
    ];
  }
};
</script>

<style lang="scss" scoped>
.list {
  display: flex;
  flex-direction: column;
  min-width: 250px;
  height: 100%;
  border-right: solid 1px #ccc;
  overflow: auto;
  flex-shrink: 0;

  .draggable {
    cursor: move;
    cursor: grab;
  }
}

.last {
  flex-grow: 1;
}
</style>
