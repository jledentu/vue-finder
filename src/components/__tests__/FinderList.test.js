import { vi, describe, it, beforeEach, expect } from "vitest";
import { mount } from "@vue/test-utils";
import TreeModel from "@/utils/tree-model";
import FinderList from "../FinderList.vue";

vi.mock("@/utils/tree-model");

describe("FinderList", () => {
  let treeModel;
  const tree = {
    id: "test1",
    children: [
      {
        id: "test11",
        label: "Test 11",
        selected: true,
        children: [
          {
            id: "test111",
            label: "Test 111",
          },
          {
            id: "test112",
            label: "Test 112",
          },
        ],
        isLeaf: false,
      },
      {
        id: "test12",
        label: "Test 12",
        isLeaf: true,
      },
    ],
  };
  beforeEach(() => {
    treeModel = new TreeModel(tree);
  });

  it("should match snapshot", () => {
    const wrapper = mount(FinderList, {
      props: {
        treeModel,
        items: tree.children,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should match snapshot if no item", () => {
    const wrapper = mount(FinderList, {
      props: {
        treeModel,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should match snapshot if `dragEnabled` is true", () => {
    const wrapper = mount(FinderList, {
      props: {
        treeModel,
        items: tree.children,
        dragEnabled: true,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should match snapshot if `selectable` is true", () => {
    const wrapper = mount(FinderList, {
      props: {
        treeModel,
        items: tree.children,
        selectable: true,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should handle focus with arrows up/down", async () => {
    const wrapper = mount(FinderList, {
      props: {
        treeModel,
        items: tree.children,
      },
      attachTo: document.body,
    });

    const firstItem = wrapper.findAll(".item")[0];
    const secondItem = wrapper.findAll(".item")[1];

    firstItem.element.focus();
    expect(document.activeElement).toBe(firstItem.element);

    // Up doesn't have any effect since the first item is focused
    await firstItem.trigger("keydown", { key: "ArrowUp" });
    expect(document.activeElement).toBe(firstItem.element);

    // Down navigates to the second item
    await firstItem.trigger("keydown", { key: "ArrowDown" });
    expect(document.activeElement).toBe(secondItem.element);

    // Up navigates back to the first item
    await secondItem.trigger("keydown", { key: "ArrowUp" });
    expect(document.activeElement).toBe(firstItem.element);
  });

  it("should handle focus with arrows up/down when drag enabled", async () => {
    const wrapper = mount(FinderList, {
      props: {
        treeModel,
        items: tree.children,
        dragEnabled: true,
      },
      attachTo: document.body,
    });

    const firstItem = wrapper.findAll(".item")[0];
    const secondItem = wrapper.findAll(".item")[1];

    firstItem.element.focus();
    expect(document.activeElement).toBe(firstItem.element);

    // Up doesn't have any effect since the first item is focused
    await firstItem.trigger("keydown", { key: "ArrowUp" });
    expect(document.activeElement).toBe(firstItem.element);

    // Down navigates to the second item
    await firstItem.trigger("keydown", { key: "ArrowDown" });
    expect(document.activeElement).toBe(secondItem.element);

    // Up navigates back to the first item
    await secondItem.trigger("keydown", { key: "ArrowUp" });
    expect(document.activeElement).toBe(firstItem.element);
  });
});
