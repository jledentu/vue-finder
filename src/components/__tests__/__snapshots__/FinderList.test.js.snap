// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinderList should match snapshot 1`] = `
<div class="list">
  <div role="button" draggable="false" class="item" tabindex="0">
    <!---->
    <!---->
    <div item="[object Object]" class="inner-item">Test 11</div>
    <div class="arrow"></div>
  </div>
  <div role="button" draggable="false" class="item" tabindex="-1">
    <!---->
    <!---->
    <div item="[object Object]" class="inner-item">Test 12</div>
    <!---->
  </div>
</div>
`;

exports[`FinderList should match snapshot if \`dragEnabled\` is true 1`] = `
<div class="list">
  <div class="drop-zone"></div>
  <div role="button" draggable="true" class="item draggable" tabindex="0">
    <!---->
    <!---->
    <div item="[object Object]" class="inner-item">Test 11</div>
    <div class="arrow"></div>
  </div>
  <div class="drop-zone"></div>
  <div role="button" draggable="true" class="item draggable" tabindex="-1">
    <!---->
    <!---->
    <div item="[object Object]" class="inner-item">Test 12</div>
    <!---->
  </div>
  <div class="drop-zone last"></div>
</div>
`;

exports[`FinderList should match snapshot if \`selectable\` is true 1`] = `
<div class="list">
  <div role="button" draggable="false" class="item" tabindex="0">
    <!----> <input type="checkbox" aria-label="Test 11">
    <div item="[object Object]" class="inner-item">Test 11</div>
    <div class="arrow"></div>
  </div>
  <div role="button" draggable="false" class="item" tabindex="-1">
    <!----> <input type="checkbox" aria-label="Test 12">
    <div item="[object Object]" class="inner-item">Test 12</div>
    <!---->
  </div>
</div>
`;

exports[`FinderList should match snapshot if no item 1`] = `<div class="list"></div>`;
